services/upgrader/upgrader.go

```diff
@@ -0,0 +1,250 @@
+		return errors.New("You can't upgrade your code to enterprise because you are already in enterprise code")
+	}
+	return nil
+}
+
+func UpgradeToE0() error {
+	if !atomic.CompareAndSwapInt32(&upgrading, 0, 1) {
+		mlog.Warn("Trying to upgrade while other upgrade is running")
+		return errors.New("One upgrade is already running.")
+	}
+	defer atomic.CompareAndSwapInt32(&upgrading, 1, 0)
+
+	upgradePercentage = 1
+	upgradeError = nil
+
+	executablePath, err := os.Executable()
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = errors.New("error getting the executable path")
+		mlog.Error("Unable to get the mattermost executable path", mlog.Err(err))
+		return err
+	}
+
+	filename, err := download(getCurrentVersionTgzUrl())
+	if err != nil {
+		if filename != "" {
+			os.Remove(filename)
+		}
+		upgradeError = fmt.Errorf("error downloading the new version (percentage: %d)", upgradePercentage)
+		mlog.Error("Unable to download the mattermost server version", mlog.Int64("percentage", upgradePercentage), mlog.String("url", getCurrentVersionTgzUrl()), mlog.Err(err))
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(filename)
+
+	err = extractBinary(executablePath, filename)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = err
+		mlog.Error("Unable to extract the binary from the downloaded file", mlog.Err(err))
+		return err
+	}
+	upgradePercentage = 100
+	return nil
+}
+
+func UpgradeToE0Status() (int64, error) {
+	return upgradePercentage, upgradeError
+}
+
+func download(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	out, err := ioutil.TempFile("", "*_mattermost.tar.gz")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	counter := &writeCounter{total: resp.ContentLength}
+	_, err = io.Copy(out, io.TeeReader(resp.Body, counter))
+	return out.Name(), err
+}
+
+func extractBinary(executablePath string, filename string) error {
+	gzipStream, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for {
+		header, err := tarReader.Next()
+
+		if err == io.EOF {
+			return errors.New("Unable to find mattermost binary in the downloaded version")
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if header.Typeflag == tar.TypeReg && header.Name == "mattermost/bin/mattermost" {
+			os.Rename(executablePath, executablePath+".bak")
+			outFile, err := os.Create(executablePath)
+			if err != nil {
+				os.Rename(executablePath+".bak", executablePath)
+				return err
+			}
+			defer outFile.Close()
+			if _, err := io.Copy(outFile, tarReader); err != nil {
+				os.Remove(executablePath)
+				os.Rename(executablePath+".bak", executablePath)
+				return err
+			}
+			os.Remove(executablePath + ".bak")
+			os.Chmod(executablePath, 0755)
```

Comment:
System administrators may use a different config than `0755` for their binary, can we read the permissions before and restore it for the new file?

Reply:
Yes, sure.

Reply:
Actually, this should have the right permissions from the tar file.